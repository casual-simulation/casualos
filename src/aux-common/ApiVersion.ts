/**
 * Gets the API version that the AUX common library currently supports.
 * This needs to get manually incremented whenever a new API version is added.
 * Basically a really barebones start for API versioning.
 */
export const apiVersion = 2;
