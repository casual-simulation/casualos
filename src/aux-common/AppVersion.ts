/* CasualOS is a set of web-based tools designed to facilitate the creation of real-time, multi-user, context-aware interactive experiences.
 *
 * Copyright (c) 2019-2025 Casual Simulation, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

export interface AppVersion {
    /**
     * Gets the git tag that this version was built from.
     */
    gitTag: string;

    /**
     * Gets the Git hash that this version was built from.
     */
    gitHash: string;

    /**
     * Gets the API that this app has support for.
     */
    apiVersion: number;
}
