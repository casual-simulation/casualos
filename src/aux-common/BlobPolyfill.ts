if (!globalThis.Blob) {
    if (typeof process === 'undefined') {
        console.warn(
            '[BlobPolyfill] Loading Blob stub! This should not happen in browser environments!'
        );
    }
    (<any>globalThis).Blob = class {
        type: string;
        parts: any[];

        constructor(parts: any[], options: any) {
            this.parts = parts.slice();
            this.type = options?.type;
        }

        async arrayBuffer(): Promise<ArrayBuffer> {
            let bytes = [] as number[];
            const encoder = new TextEncoder();
            for (const part of this.parts) {
                if (typeof part === 'string') {
                    const encoded = encoder.encode(part);
                    bytes.push(...encoded);
                } else if (part instanceof ArrayBuffer) {
                    bytes.push(...new Uint8Array(part));
                } else if (ArrayBuffer.isView(part)) {
                    bytes.push(...new Uint8Array(part.buffer));
                } else if (part instanceof Blob) {
                    const buffer = await part.arrayBuffer();
                    bytes.push(...new Uint8Array(buffer));
                }
            }

            const view = new Uint8Array(bytes);
            return view.buffer;
        }
    };
}
