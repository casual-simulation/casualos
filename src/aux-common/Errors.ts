/* CasualOS is a set of web-based tools designed to facilitate the creation of real-time, multi-user, context-aware interactive experiences.
 *
 * Copyright (c) 2019-2025 Casual Simulation, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

export type ServerError = 'server_error';

/**
 * Defines an error that occurs when the user is not logged in but they are required to be in order to perform an action.
 */
export type NotLoggedInError = 'not_logged_in';

/**
 * Defines an error that occurs when the user does not have the right permissions to perform an action.
 */
export type NotAuthorizedError = 'not_authorized';

/**
 * Defines an error that occurrs when the user tries to perform an action that is not supported by the current deployment.
 */
export type NotSupportedError = 'not_supported';

/**
 * Defines an error that occurrs when the user tries to perform an action that requires them to be subscribed.
 */
export type NotSubscribedError = 'not_subscribed';

/**
 * Defines an error that occurrs when the user tries to perform an action that requires a subscription tier that .
 */
export type InvalidSubscriptionTierError = 'invalid_subscription_tier';

/**
 * Defines an error that occurs when the user tries to perform an action that would exceed their subscription limit.
 */
export type SubscriptionLimitReached = 'subscription_limit_reached';
