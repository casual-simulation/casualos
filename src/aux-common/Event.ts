// [Ryan] I would love to have multiple versions of Event and Listener that support more than one generic parameter.
// However, this is not supported by JavaScript. TypeScript's compiler complains about the class name already being declared even though
// the generic parameters are different. This is because when transpiled to JavaScript,all generic types are stripped.
// So the best way to handle this is if you need mutiple objects to store them in a custom JavaScript object / Typescript interface.

/**
 * Event that has no arguments.
 */
export class Event {
    private _listeners: EventListener[] = [];

    public addListener(listener: EventListener) {
        this._listeners.push(listener);
    }

    public removeListener(listener: EventListener) {
        const index = this._listeners.indexOf(listener);
        if (index !== -1) {
            this._listeners.splice(index, 1);
        }
    }

    public removeAllListeners() {
        this._listeners = [];
    }

    public invoke() {
        this._listeners.forEach((l) => {
            l();
        });
    }
}

/**
 * Signature of event listener that has no arguments.
 */
declare type EventListener = () => void;

/**
 * Event that takes typed argument.
 */
export class ArgEvent<T> {
    private _listeners: ArgEventListener<T>[] = [];

    public addListener(listener: ArgEventListener<T>) {
        this._listeners.push(listener);
    }

    public removeListener(listener: ArgEventListener<T>) {
        const index = this._listeners.indexOf(listener);
        if (index !== -1) {
            this._listeners.splice(index, 1);
        }
    }

    public removeAllListeners() {
        this._listeners = [];
    }

    public invoke(arg: T) {
        this._listeners.forEach((l) => {
            l(arg);
        });
    }
}

/**
 * Signature of event listener that accepts typed argument.
 */
declare type ArgEventListener<T> = (arg: T) => void;
