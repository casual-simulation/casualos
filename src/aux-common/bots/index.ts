export * from './Bot';
export * from './BotEvents';
export * from './BotCalculationContext';
export * from './BotCalculationContextFactory';
export * from './BotCalculations';
export * from './BotActions';
export * from './TagUpdatedEvent';
export * from './BotIndex';
export * from './BotLookupTable';
export * from './BotLookupTableHelper';
export * from './StateUpdatedEvent';
export * from './StoredAux';
export * from './AuxStateHelpers';
