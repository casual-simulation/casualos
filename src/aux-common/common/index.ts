export * from './Action';
export * from './ConnectionInfo';
export * from './RemoteActions';
export * from './ConnectionIndicator';
export * from './StatusUpdate';
export * from './CurrentVersion';
export * from './StatusUpdateUtils';
export * from './LoadingProgress';
export * from './ConnectionToken';
export * from './Iterators';
export * from './PolicyPermissions';
export * from './DenialReason';
