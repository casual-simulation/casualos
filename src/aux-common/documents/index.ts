export * from './SharedDocument';
export * from './SharedDocumentConfig';
export * from './SharedDocumentFactories';
export * from './YjsSharedDocument';
export * from './RemoteYjsSharedDocument';
