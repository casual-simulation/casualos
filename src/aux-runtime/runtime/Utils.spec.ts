import {
    embedBase64InPdf,
    formatAuthToken,
    fromHexString,
    getEmbeddedBase64FromPdf,
    isPromise,
    parseAuthToken,
    toHexString,
} from './Utils';
import '@casual-simulation/aux-common/BlobPolyfill';

describe('embedBase64InPdf()', () => {
    it('should reference the given data in the PDF', () => {
        const data = 'abcdefghiabcdefghi';
        const result = embedBase64InPdf(data);

        expect(result).toContain(data);
        expect(result).toMatchSnapshot();
    });
});

describe('getEmbeddedBase64FromPdf()', () => {
    it('should return the data that was embedded in the PDF', () => {
        const data = 'abcdefghiabcdefghi';
        const pdf = embedBase64InPdf(data);
        const result = getEmbeddedBase64FromPdf(pdf);

        expect(result).toEqual(data);
    });
});

describe('formatAuthToken()', () => {
    const cases = [['myToken', 'myService', 'myToken.myService']];

    it.each(cases)('should format %s and %s', (token, service, expected) => {
        const result = formatAuthToken(token, service);

        expect(result).toBe(expected);
    });
});

describe('parseAuthToken()', () => {
    const cases = [
        ['myToken.myService', ['myToken', 'myService']] as const,
        ['myToken.mySer.vice', ['myToken', 'mySer.vice']] as const,
        ['myToken', null as any] as const,
    ] as const;

    it.each(cases)('should format %s and %s', (token, expected) => {
        const result = parseAuthToken(token);

        expect(result).toEqual(expected);
    });
});

describe('fromHexString()', () => {
    const cases: [string, number][] = [
        ['00', 0],
        ['01', 1],
        ['02', 2],
        ['03', 3],
        ['04', 4],
        ['05', 5],
        ['06', 6],
        ['07', 7],
        ['08', 8],
        ['09', 9],
        ['0A', 10],
        ['0B', 11],
        ['0C', 12],
        ['0D', 13],
        ['0E', 14],
        ['0F', 15],

        ['10', 16],
        ['20', 32],
        ['30', 48],
        ['40', 64],
        ['50', 80],
        ['60', 96],
        ['70', 112],
        ['80', 128],
        ['90', 144],
        ['A0', 160],
        ['B0', 176],
        ['C0', 192],
        ['D0', 208],
        ['E0', 224],
        ['F0', 240],
        ['FF', 255],
    ];

    it.each(cases)('should parse %s to %s', (given, expected) => {
        const array = fromHexString(given);
        expect(array).toEqual(new Uint8Array([expected]));
    });

    it('should support long hex strings', () => {
        expect(fromHexString('abcdef1230')).toEqual(
            new Uint8Array([171, 205, 239, 18, 48])
        );
        expect(fromHexString('FFFEFD')).toEqual(
            new Uint8Array([255, 254, 253])
        );
    });
});

describe('toHexString()', () => {
    const cases: [number, string][] = [
        [0, '00'],
        [1, '01'],
        [2, '02'],
        [3, '03'],
        [4, '04'],
        [5, '05'],
        [6, '06'],
        [7, '07'],
        [8, '08'],
        [9, '09'],
        [10, '0a'],
        [11, '0b'],
        [12, '0c'],
        [13, '0d'],
        [14, '0e'],
        [15, '0f'],

        [16, '10'],
        [32, '20'],
        [48, '30'],
        [64, '40'],
        [80, '50'],
        [96, '60'],
        [112, '70'],
        [128, '80'],
        [144, '90'],
        [160, 'a0'],
        [176, 'b0'],
        [192, 'c0'],
        [208, 'd0'],
        [224, 'e0'],
        [240, 'f0'],
        [255, 'ff'],
    ];

    it.each(cases)('should transform %d to %s', (given, expected) => {
        const str = toHexString(new Uint8Array([given]));
        expect(str).toBe(expected);
    });

    it('should support long hex strings', () => {
        expect(toHexString(new Uint8Array([171, 205, 239, 18, 48]))).toBe(
            'abcdef1230'
        );
        expect(toHexString(new Uint8Array([255, 254, 253]))).toBe('fffefd');
    });
});

describe('isPromise()', () => {
    it('should return true if the value is a promise', () => {
        let p = new Promise((resolve, reject) => {});
        expect(isPromise(p)).toBe(true);
    });

    it('should return false if the value is not a promise', () => {
        expect(isPromise({})).toBe(false);
        expect(
            isPromise({
                then: () => {},
                catch: () => {},
            })
        ).toBe(false);
        expect(isPromise(null)).toBe(false);
        expect(isPromise(undefined)).toBe(false);
        expect(isPromise(123)).toBe(false);
        expect(isPromise('abc')).toBe(false);
        expect(isPromise(true)).toBe(false);
    });
});
