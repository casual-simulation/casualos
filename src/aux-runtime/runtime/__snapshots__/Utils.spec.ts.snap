// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`embedBase64InPdf() should reference the given data in the PDF 1`] = `
"%PDF-1.3
%����

1 0 obj
<<
/Type /Catalog
/Outlines 2 0 R
/Pages 3 0 R
>>
endobj

2 0 obj
<<
/Type /Outlines
/Count 0
>>
endobj

3 0 obj
<<
/Type /Pages
/Count 1
/Kids [ 4 0 R ] 
>>
endobj

4 0 obj
<<
/Type /Page
/Parent 3 0 R
/Resources <<
/Font <<
/F1 9 0 R 
>>
/ProcSet 8 0 R
>>
/MediaBox [0 0 512 512]
/Annots [ 25 0 R ]
/Contents 5 0 R
>>
endobj

5 0 obj
<< /Length 1074 >>
stream
BT
/F1 28 Tf
183 412 Td
( CasualOS ) Tj
ET
BT
/F1 16 Tf
150 256 Td
( Open) Tj
ET
BT
0 0 1 rg
0 0 1 RG
/F1 16 Tf
193.25 256 Td
( https://casualos.com ) Tj
ET
endstream
endobj

8 0 obj
[/PDF /Text]
endobj

9 0 obj
<<
/Type /Font
/Subtype /Type1
/Name /F1
/BaseFont /Helvetica
/Encoding /WinAnsiEncoding
>>
endobj

10 0 obj
<<
/Creator (Casual Simulation)
/Producer (CasualOS)
/CreationDate (D:20210720000000)
>>
endobj

25 0 obj                                                % object ID 26 which is referenced by the OBJR in Object 11
<<
/Type /Annot
/Subtype /Link
/Border [0 0 0]                                       %a colorless border
/Rect [193.25 245 350 276]                              %the boundaries defining target area where link annotation is active
/A <</S /URI /URI (https://casualos.com)>>              % The action that the link should perform (Go to casualos.com)
>>
endobj

9999909 0 obj
 (abcdefghiabcdefghi)
endobj

xref
0 11
0000000000 65535 f
0000000019 00000 n
0000000093 00000 n
0000000147 00000 n
0000000222 00000 n
0000000390 00000 n
0000001522 00000 n
0000001690 00000 n
0000002423 00000 n
0000002456 00000 n
0000002574 00000 n

trailer
<<
/Size 11
/Root 1 0 R
/Info 10 0 R
>>

startxref
2714
%%EOF
"
`;
